### ISSUE TITLE
client_id not in aud when using keycloak gatekeeper

### ISSUE THREAD

**Issue by abstractj** [keycloak/keycloak-gatekeeper#567]
## What: 

  When using the latest 4.6.0 version of Keycloak i am not able to login via the Keycloak Gatekeeper proxy.
  
  {code}
  unable to verify the id token	{"error": "oidc: JWT claims invalid: invalid claims, cannot find 'client_id' in 'aud' claim, aud=[master-realm account], client_id=foo_test"}
  {code}
  
  Previous versions worked out of the box (i tested 4.4.0 and 4.5.0).

## Reference:

- [KEYCLOAK-8954 - client_id not in aud when using keycloak gatekeeper](https://issues.redhat.com/browse/KEYCLOAK-8954)

**Comment by abstractj** [comment]
@sirianni could you please be more specific? Are you talking about https://github.com/louketo/louketo-proxy/pull/538 ?

**Comment by jangaraj** [comment]
My guess: https://stackoverflow.com/questions/53550321/keycloak-gatekeeper-aud-claim-and-client-id-do-not-match

**Comment by sirianni** [comment]
> @sirianni could you please be more specific? Are you talking about #538 ?

Yes, I checked out and compiled your branch to work around https://github.com/coreos/go-oidc/issues/182 but then hit this issue.  I'm just pointing out that the same issue applies when using Okta as an IdP and not just "the 4.6.0 version of Keycloak"

I'm not sure why the code is trying to verify an access token using the OIDC library.  The access token does not even need to be a JWT.  It could be an opaque token.  https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowTokenValidation

**Comment by ackerleytng** [comment]
@patriksletmo also noted that gatekeeper is verifying the access token as if it is an id token here https://lists.jboss.org/pipermail/keycloak-dev/2019-October/012699.html

**Comment by HansK-p** [comment]
I assume it's well known that something happened with Keycloak 4.6.0 and the audience field. Ref. https://stackoverflow.com/questions/53550321/keycloak-gatekeeper-aud-claim-and-client-id-do-not-match. Based on comments above I'd guess there is more to this bug report than only the change introduced in Keycloak 4.6.0.

The link has a description as to how the client_id can be added to the audience field after an upgrade to Keycloak 4.6.0+ (including the latest version which is 10.0.1).

**Comment by nxadm** [comment]
> Based on comments above I'd guess there is more to this bug report than only the change introduced in Keycloak 4.6.0.

As far as I can see (and how the workaround works), the bug is not in Keycloak and other providers but in louketo as it handles the response.

**Comment by abstractj** [comment]
That's correct, the bug is not in Keycloak and has to be fixed in Louketo. To fix this, we've been working to transition from go-oidc v1 to go-oidc v2, which contains several bugs. You can keep track of this by watching the following issues:
- #630
- #519
- #517
The initial work to migrate oauth2 is [here](https://github.com/louketo/louketo-proxy/pull/630) for those interested in giving it a try.

**Comment by adpande** [comment]
I am also facing the same issue.

**Comment by joao-dantas** [comment]
Hi, 
Is there any plan to fix this issue?

**Comment by CuBiC3D** [comment]
@joao-dantas see #683
